File: native/Sources/Scribe/History.swift
import Foundation

/// A single completed dictation, held only in memory (privacy: never persisted).
struct DictationRecord {
    let raw: String
    let final: String
    let engine: String
    let cleaned: Bool
    let at: Date
    let durationMs: Int
}

/// Bounded, thread-safe in-memory dictation history (privacy: never persisted).
final class History {
    private let maxLen: Int
    private var storage: [DictationRecord] = []
    private let lock = NSLock()

    init(maxLen: Int) {
        self.maxLen = maxLen
    }

    func append(_ r: DictationRecord) {
        lock.lock()
        defer { lock.unlock() }
        storage.append(r)
        if storage.count > maxLen {
            storage.removeFirst(storage.count - maxLen)
        }
    }

    /// Newest first.
    func items() -> [DictationRecord] {
        lock.lock()
        defer { lock.unlock() }
        return Array(storage.reversed())
    }
}

File: src/susurro/history.py
"""Bounded, thread-safe in-memory dictation history (privacy: never persisted)."""

from __future__ import annotations

import threading
from collections import deque
from dataclasses import dataclass


@dataclass(frozen=True)
class Record:
    raw: str
    final: str
    engine: str
    cleaned: bool
    at: float
    duration_ms: int


class History:
    def __init__(self, maxlen: int):
        self._items: deque[Record] = deque(maxlen=maxlen)
        self._lock = threading.Lock()

    def append(self, record: Record) -> None:
        with self._lock:
            self._items.append(record)

    def items(self) -> list[Record]:
        with self._lock:
            return list(reversed(self._items))
